Use Maven Wrapper

rahulsom opened the pull request:
This reduces the number of dependencies for development.
Also, it enables everyone to use the same version of maven.

### Proposed changelog entries

* Developer: Use maven wrapper to provide consistent maven version for development builds

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

daniel-beck reviewed (dismissed):
All of the unnecessary whitespace changes need to be reverted.

oleg-nenashev commented:
Before the time is spent on reverting whitespaces, it would be great to
discuss the feasibility. IMHO it needs more feedback and justification.

Personally I do not see much added value for development needs in this
wrapper. If we were going down this road, I would have rather preferred a
Makefile with common developer use-cases


On Sat, May 23, 2020, 09:41 Daniel Beck <notifications@github.com> wrote:

> *@daniel-beck* requested changes on this pull request.
>
> All of the unnecessary whitespace changes need to be reverted.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkinsci/jenkins/pull/4736#pullrequestreview-417258902>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAW4RIBI6LSMRVQW7MAHMLDRS54ZBANCNFSM4NIGJCGQ>
> .
>

oleg-nenashev commented:
Somewhat related: https://twitter.com/rfscholte/status/1264112670158061568?s=19

timja commented on .mvn/wrapper/maven-wrapper.properties:
```diff
@@ -0,0 +1,2 @@
+distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.5.4/apache-maven-3.5.4-bin.zip
```
we use 3.6.3 on our infrastructure currently

```suggestion
distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.6.3/apache-maven-3.6.3-bin.zip
```

rahulsom replied:
The docs recommended 3.5.4 or above.
@oleg-nenashev This is an example of the kind of problem a wrapper solves.

timja commented on core/src/main/java/hudson/init/InitStrategy.java:
```diff
@@ -30,7 +30,7 @@
  * <p>
  * To register, put {@link MetaInfServices} on your implementation.
- * 
+ *
```
no need change this file at all

timja commented on core/src/main/java/hudson/model/Slave.java:
```diff
@@ -438,7 +438,7 @@ public URL getURL() throws IOException {
         }
 
-        /** Useful for {@code JenkinsRule.createSlave}, {@code mvn jetty:run}, etc. */
+        /** Useful for {@code JenkinsRule.createSlave}, {@code ./mvnw jetty:run}, etc. */
```
or this

timja commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -5149,7 +5149,7 @@ private static void computeVersion(ServletContext context) {
     /**
      * Parses {@link #VERSION} into {@link VersionNumber}, or null if it's not parseable as a version number
-     * (such as when Jenkins is run with {@code mvn jetty:run})
+     * (such as when Jenkins is run with {@code ./mvnw jetty:run})
```
or this

timja commented on test/src/test/java/jenkins/agents/WebSocketAgentsTest.java:
```diff
@@ -76,7 +76,7 @@
     /**
      * Verify basic functionality of an agent in {@code -webSocket} mode.
-     * Requires {@code remoting} to have been {@code mvn install}ed.
+     * Requires {@code remoting} to have been {@code ./mvnw install}ed.
```
or this

timja reviewed (commented):
I don't think we need to change any of the java source files 

good to see that the maven wrapper is now merged in to maven (although unreleased https://issues.apache.org/jira/browse/MNG-5937)

rahulsom committed:
Use Maven Wrapper

This reduces the number of dependencies for development.
Also, it enables everyone to use the same version of maven.

rahulsom committed:
Bump maven version to 3.6.3

Co-authored-by: Tim Jacomb <t.jacomb@kainos.com>

rahulsom commented:
I can hold off on this change until Maven releases the wrapper.

timja commented on core/src/main/java/hudson/Main.java:
```diff
@@ -220,8 +220,8 @@ private static HttpURLConnection open(URL url) throws IOException {
     /**
-     * Set to true if we are running inside {@code mvn jetty:run}.
-     * This is also set if running inside {@code mvn hpi:run} since plugins parent POM 2.30.
+     * Set to true if we are running inside {@code ./mvnw jetty:run}.
```
Revert this also please

rahulsom committed:
Revert mvn -> mvnw changes in javadoc

oleg-nenashev commented:
@jenkinsci/core Any opinions about merging the PR? Maven wrapper is officially in GA

timja reviewed (approved):
+1, always better to not have to deal with build inconsistencies from different versions, this allows to only support 1 version.

MarkEWaite reviewed (approved):
I confirmed this is well-behaved on OpenBSD.  No objections from me.

jglick commented on Jenkinsfile:
```diff
@@ -43,7 +43,7 @@ for(j = 0; j < jdks.size(); j++) {
                             // Actually run Maven!
                             // -Dmaven.repo.local=… tells Maven to create a subdir in the temporary directory for the local Maven repository
-                            def mvnCmd = "mvn -Pdebug -U -Dset.changelist help:evaluate -Dexpression=changelist -Doutput=$changelistF clean install ${runTests ? '-Dmaven.test.failure.ignore' : '-DskipTests'} -V -B -ntp -Dmaven.repo.local=$m2repo -e"
+                            def mvnCmd = "./mvnw -Pdebug -U -Dset.changelist help:evaluate -Dexpression=changelist -Doutput=$changelistF clean install ${runTests ? '-Dmaven.test.failure.ignore' : '-DskipTests'} -V -B -ntp -Dmaven.repo.local=$m2repo -e"
```
Note that we are now using ACI agents which bundle Maven, yet not using the bundled version. Deserves a comment above (L27) to avoid confusion.

rahulsom replied:
Done.

jglick commented on Jenkinsfile:
```diff
@@ -87,7 +87,7 @@ builds.ath = {
             checkout scm
             withMavenEnv(["JAVA_OPTS=-Xmx1536m -Xms512m",
```
Again potentially misleading.

rahulsom replied:
That is now called withJdkEnv and only configures the JDK for non-linux environments.

jglick commented on mvnw:
```diff
@@ -0,0 +1,310 @@
+
+##########################################################################################
+# Extension to allow automatically downloading the maven-wrapper.jar from Maven-central
+# This allows using the maven wrapper in projects that prohibit checking in binary data.
```
Which, IMO, we should, and remove `.mvn/wrapper/maven-wrapper.jar` from this PR. (I guess that would require an addition to `.gitignore`?)

rahulsom replied:
Done.

jglick commented on war/pom.xml:
```diff
@@ -514,7 +514,7 @@ THE SOFTWARE.
               <!-- enable the plugins in main by default -->
               <name>hudson.bundled.plugins</name>
-              <value><!-- run "mvn install" once will generate the.hpl -->
+              <value><!-- run "./mvnw install" once will generate the.hpl -->
```
or 
```suggestion
              <value><!-- run "../mvnw install" once will generate the.hpl -->
```

jglick commented on core/pom.xml:
```diff
@@ -828,7 +828,7 @@ THE SOFTWARE.
     </profile>
     <profile>
-      <!-- Run SpotBugs for better error detection. Run as "mvn -Pspotbugs install site". -->
+      <!-- Run SpotBugs for better error detection. Run as "./mvnw -Pspotbugs install site". -->
```
perhaps
```suggestion
      <!-- Run SpotBugs for better error detection. Run as "../mvnw -Pspotbugs install site". -->
```

(not sure what was meant by this comment)

jglick reviewed (changes requested):
I think I am uncomfortable with the JAR in the source tree. Seems unnecessary and a supply chain attack risk.

rahulsom committed:
Update war/pom.xml

Co-authored-by: Jesse Glick <jglick@cloudbees.com>

rahulsom committed:
Exclude jar from git

jglick commented on Jenkinsfile:
```diff
@@ -24,6 +24,8 @@ for(j = 0; j < jdks.size(); j++) {
     def jdk = jdks[j]
     builds["${buildType}-jdk${jdk}"] = {
         // see https://github.com/jenkins-infra/documentation/blob/master/ci.adoc#node-labels for information on what node types are available
+		// We do not rely on the maven version provided by the agent. The maven version is selected by the wrapper in
+		// .mvn/wrapper/maven-wrapper.properties
```
```suggestion
        // We do not rely on the maven version provided by the agent. The maven version is selected by the wrapper in
        // .mvn/wrapper/maven-wrapper.properties
```

jglick commented on Jenkinsfile:
```diff
@@ -39,8 +41,8 @@ for(j = 0; j < jdks.size(); j++) {
-                        withMavenEnv(["JAVA_OPTS=-Xmx1536m -Xms512m",
-                                    "MAVEN_OPTS=-Xmx1536m -Xms512m"], buildType, jdk) {
+                        withJdkEnv(["JAVA_OPTS=-Xmx1536m -Xms512m",
+									"MAVEN_OPTS=-Xmx1536m -Xms512m"], buildType, jdk) {
```
```suggestion
                                    "MAVEN_OPTS=-Xmx1536m -Xms512m"], buildType, jdk) {
```

jglick commented on Jenkinsfile:
```diff
@@ -85,8 +87,8 @@ builds.ath = {
-            withMavenEnv(["JAVA_OPTS=-Xmx1536m -Xms512m",
-                          "MAVEN_OPTS=-Xmx1536m -Xms512m"], 8) {
+            withJdkEnv(["JAVA_OPTS=-Xmx1536m -Xms512m",
+						"MAVEN_OPTS=-Xmx1536m -Xms512m"], 8) {
```
```suggestion
                          "MAVEN_OPTS=-Xmx1536m -Xms512m"], 8) {
```

jglick reviewed (approved):
Looks better, thanks!

oleg-nenashev reviewed (commented):
I am not 100% sure about changing Jenkinsfile, because it limits our ability to manage environment from the Pipeline Library. IIUC it will now also require Maven download on every build after the patch, right?

timja commented:
> I am not 100% sure about changing Jenkinsfile, because it limits our ability to manage environment from the Pipeline Library. IIUC it will now also require Maven download on every build after the patch, right?

it's not managed from the pipeline library afaik, it just loads the mvn tool.

It should be loaded once per agent in the current state.

jglick commented:
> once per agent

Which means every build, at least for the Linux branch, since the agents are ACI and thus not reused.

jglick commented on .mvn/wrapper/maven-wrapper.properties:
```diff
@@ -0,0 +1,2 @@
+distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.6.3/apache-maven-3.6.3-bin.zip
```
```suggestion
distributionUrl=https://repo.jenkins-ci.org/public/org/apache/maven/apache-maven/3.6.3/apache-maven-3.6.3-bin.zip
```

rahulsom committed:
Yse jenkins-ci mirror for downloading maven

Co-authored-by: Jesse Glick <jglick@cloudbees.com>

jtnord commented on Dockerfile:
```diff
@@ -17,7 +17,7 @@ COPY licenseCompleter.groovy /jenkins/src/licenseCompleter.groovy
 
 WORKDIR /jenkins/src/
-RUN mvn clean install --batch-mode -Plight-test
+RUN ./mvnw clean install --batch-mode -Plight-test
```
this seems wrong.  if we are using whatever version of maven then do not use a container with a specific version of maven embedded.
![image](https://user-images.githubusercontent.com/494726/96262667-03fa4880-0fba-11eb-89ee-951e5f946aed.png)

jglick replied:
Right, should change the base image here.

jtnord reviewed (commented):
I wonder if this should be merged as is as keeping the files up to date will be a chore.
also downlaoding maven in the CI system as opposed to obtaining it from the ~master~ controller (or what is in the agent image) may increase ingress costs.

there is a plugin to configure the wrapper and keep it up-to-date (which apparently will be introduced in maven 3.7.0 and may use a substantially different techinque?)

https://maven.apache.org/plugins/maven-wrapper-plugin/usage.html

jglick commented:
> downlaoding maven in the CI system

Not a problem I think because we are now getting it from Artifactory, like all of our deps.

alecharp commented:
@rahulsom Could you please review the comments and address the merge conflicts? Thanks.

daniel-beck commented:
> Not a problem I think because we are now getting it from Artifactory, like all of our deps.

Artifactory is hosted outside our other infra, and INFRA-2754 says we shouldn't burden it even more.

alecharp commented on Dockerfile:
```diff
@@ -1,7 +1,7 @@
-FROM maven:3.5.4-jdk-8 as builder
+FROM openjdk:8-jdk as builder
 
 COPY .mvn/ /jenkins/src/.mvn/
```
```suggestion
COPY .mvn/ /jenkins/src/.mvn/
COPY mvnw /jenkins/src/mvnw
```

timja replied:
tbh I don't think this Dockerfile is used and should prob just be removed, no one is maintaining it.

alecharp reviewed (changes requested):
There is a problem in the Dockerfile
> COPY test-jdk8/ /jenkins/src/test-jdk8/

needs to be removed

alecharp commented:
@rahulsom could you please attend to the merge conflicts so we can move this PR forward? Thanks.

NotMyFault reviewed (commented):
If we process this PR further, I'd vouch for regenerating the wrapper files with the appropriate command (`mvn wrapper:wrapper -Dmaven=3.8.5`) to warrant we initially ship up to date files, further updates, if needed, will be handled by dependabot.

timja commented on .mvn/wrapper/maven-wrapper.properties:
```diff
@@ -0,0 +1,18 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.8.5/apache-maven-3.8.5-bin.zip
```
we can't use 3.8.5, we're hit by https://issues.apache.org/jira/browse/MNG-7432

NotMyFault replied:
Oh, I wasn't aware of the open issue when I commented on the PR last time. I guess we can safely use 3.8.1 as recommended in the contributing.md.

timja replied:
3.8.4 is what's used on CI

jglick replied:
Of course this kind of regression is a prime argument for using the Maven wrapper!

jglick commented:
Given that `mvnd` is so much nicer than `mvn` especially on larger repos like this one, I wonder if we should not just put this on hold until it becomes an official part of Maven and supported by the wrapper. If it ever does; `mvnd` includes a GraalVM-based native image which you would not want to commit.

Anyway, if we are going to commit the wrapper to a repo, it only makes sense to do that if CI actually uses it, which https://github.com/jenkinsci/jenkins/blob/e3f8bc052d6065d3fcf6430f4c7c029c1fd25e54/Jenkinsfile#L70 currently will not AFAICT. https://github.com/jenkinsci/jenkins/blob/e3f8bc052d6065d3fcf6430f4c7c029c1fd25e54/Jenkinsfile#L37-L41 selects images with both Java and Maven. It is nice for CI to have the binary already available (though we download far more from Artifactory during the build anyway) but I would expect a proposed version change of Maven in the repo to actually be tested, or conversely that proposed POM or other code changes be tested against the pinned version of Maven.

rahulsom committed:
Remove empty Dockerfile

basil reviewed (commented):
> Anyway, if we are going to commit the wrapper to a repo, it only makes sense to do that if CI actually uses it

+1. I suggest https://github.com/jenkins-infra/pipeline-library/blob/216eee69ae367cf11871013fb52b0dfae61b73ee/vars/infra.groovy#L73-L96 be updated to support Maven Wrapper, with this PR consuming the new functionality in this repository's `Jenkinsfile`.

alecharp commented:
At this moment, do we want to wait for `mvnd` to be part of Maven (which should be in Maven 4 or Maven 5, but we don't know when this will be). 

If we have a down-stream pull request to the `Jenkinsfile` to use `./mvnw` rather than `mvn`, I'd say we are good to go with the current content, no?

basil commented:
> If we have a down-stream pull request to the `Jenkinsfile` to use `./mvnw` rather than `mvn`

But we do not, so this PR is stalled. I have added the `stalled` label pending the creation of downstream PRs. If the review feedback remains unaddressed after approximately one month, this PR will transition from the `stalled` label to the `proposed-for-close` label and closed approximately one week after that.
